### PR TITLE
added a qemu-img map an unaligned image case

### PR DESCRIPTION
**Pull request by hereischen**
id: 1653579
Signed-off-by: Haotong Chen <hachen@redhat.com>

### CODE REVIEW

File: qemu/tests/qemu_img_map_unaligned_image.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    def _generate_random_string(max_len=19):
+        """Generate random alphabets string in the range of [1, max_len]."""
+        random_str = ''.join(random.choice(
+            string.ascii_lowercase) for _ in range(random.randint(0, max_len)))
+        return random_str, len(random_str)
+
+    def _verify_qemu_img_map(output, str_len):
+        """Verify qemu-img map's output."""
+        # str_len + 1 is expected qeme-img map output for length
```
Comment (hellenmao) [inline qemu/tests/qemu_img_map_unaligned_image.py]: Maybe 'str_len + 1 is the actual bytes in storage' is better?

File: qemu/tests/qemu_img_map_unaligned_image.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        expected = {"start": 0, "length": str_len + 1, "depth": 0,
+                    "zero": False, "data": True, "offset": 0}
+        res = json.loads(output)[0]
+        if res != expected:
+            test.fail("qemu-img map on an unaligned image is not correct.")
```
Comment (hellenmao) [inline qemu/tests/qemu_img_map_unaligned_image.py]: 'The dumped metedata info of the unaligned image is not correct' would be better, thanks.

File: qemu/tests/qemu_img_map_unaligned_image.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    :param env: Dictionary with test environment
+    """
+    def _generate_random_string(max_len=19):
+        """Generate random alphabets string in the range of [1, max_len]."""
+        random_str = ''.join(random.choice(
+            string.ascii_lowercase) for _ in range(random.randint(0, max_len)))
+        return random_str, len(random_str)
+
+    def _verify_qemu_img_map(output, str_len):
+        """Verify qemu-img map's output."""
```
Comment (hellenmao) [inline qemu/tests/qemu_img_map_unaligned_image.py]: Maybe 'str_len + 1 is the actual allocated bytes in storage' is better?

File: qemu/tests/qemu_img_map_unaligned_image.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+from virttest.qemu_storage import QemuImg
+
+
+def run(test, params, env):
+    """
+    qemu-img map an unaligned image.
+
+    1.create a raw file using truncate
+    2.write data into the raw file
+    3.qeme-img map to verify map function works as expected
```
Comment (hellenmao) [inline qemu/tests/qemu_img_map_unaligned_image.py]: 'verify the dumped mete-data by 'qemu-img map'' would be better?

**Comment by hellenmao** [comment]
Ack.

**Comment by PaulYuuu** [comment]
Acked-by: Yihuang Yu <yihyu@redhat.com>

File: qemu/tests/qemu_img_map_unaligned_image.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    logging.info("Create a new file %s using truncate." % img.image_filename)
+    process.run("rm -f %s" % img.image_filename)
+    process.run("truncate -s 1G %s " % img.image_filename)
+
+    random_str, str_len = _generate_random_string()
+    logging.info("Write '%s' into the file %s." % (random_str,
+                                                   img.image_filename))
+    process.run("echo '%s' > %s" % (random_str, img.image_filename),
+                shell=True)
+    output = img.map(output="json").stdout_text
```
Comment (pingl2016) [inline qemu/tests/qemu_img_map_unaligned_image.py]: Before verify cmd_result.output, you should check whether the command is failed. In the original use case, a core dump will be generated by qemu-img, when dump the metadata of an unaligned image. Then there is no output, only stderr.

File: qemu/tests/qemu_img_map_unaligned_image.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            string.ascii_lowercase) for _ in range(random.randint(1, max_len)))
+        return random_str, len(random_str)
+
+    def _verify_qemu_img_map(output, str_len):
+        """Verify qemu-img map's output."""
+        logging.info("Verify the dumped mete-data of the unaligned image.")
+        # str_len + 1 is the actual allocated bytes in storage
+        expected = {"start": 0, "length": str_len + 1, "depth": 0,
+                    "zero": False, "data": True, "offset": 0}
+        res = json.loads(output)[0]
```
Comment (pingl2016) [inline qemu/tests/qemu_img_map_unaligned_image.py]: Only check the allocated spaces are not enough when json format is used. You should also check the left space(from 'str_len + 1' to the end) are unallocated

File: qemu/tests/qemu_img_map_unaligned_image.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    img = QemuImg(img_param, data_dir.get_data_dir(), "test")
+
+    logging.info("Create a new file %s using truncate." % img.image_filename)
+    process.run("rm -f %s" % img.image_filename)
+    process.run("truncate -s 1G %s " % img.image_filename)
+
+    random_str, str_len = _generate_random_string()
+    logging.info("Write '%s' into the file %s." % (random_str,
+                                                   img.image_filename))
+    process.run("echo '%s' > %s" % (random_str, img.image_filename),
```
Comment (pingl2016) [inline qemu/tests/qemu_img_map_unaligned_image.py]: It is better to use "echo -n"

Comment (hereischen) [reply qemu/tests/qemu_img_map_unaligned_image.py]: updated, thanks

**Pull request merged**